### PR TITLE
[Fastlane.Swift] Swift's `ConfigItem` for conflicting values into Ruby's end.

### PR DESCRIPTION
**Pull request by minuscorp**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
This is a fix for many Swift related issues with `nil` and conflicting values being sent into the Ruby side, like: #17343

### Description
This introduces a new type into the _fastlane_ Swift's system called `ConfigItem`, the ability of this `enum` to conform a whole family of `DescriptibleBy*Literal` makes the transition from the current code to the new one just transparent. With a bit of boilerplate that is not going to change into future breaking APIs anytime soon.

#### Example of new implementation:

```swift
public func setupCi(force: Bool = false,
                    provider: ConfigItem<String?> = .fastlaneDefault(nil))
{
    let forceArg = RubyCommand.Argument(name: "force", value: force, type: nil)
    let providerArg = provider.asRubyArgument(name: "provider", type: nil)
    let args = [forceArg,
                providerArg]
        .compactMap { $0 }
    let command = RubyCommand(commandID: "", methodName: "setup_ci", className: nil, args: args)
    _ = runner.executeCommand(command)
}
```

### Testing Steps

1. Install:
```ruby
gem "fastlane", :git => "https://github.com/minuscorp/fastlane.git", :branch => "swift-configitem"
```
2. Introduce two non-`nil` conflicting options -> see that it does break (as it doesn't pass the `conflict_block`).
3. Introduce any other Ruby's valid set of options into the Swift end and see that it works properly.

### CODE REVIEW

File: fastlane/lib/fastlane/swift_fastlane_function.rb [hunk 1]
```diff
@@ -271,12 +279,19 @@ def build_argument_list
...
         sanitized_name = camel_case_lower(string: name)
         sanitized_name = sanitize_reserved_word(word: sanitized_name)
         type_string = type_override == :string_callback ? ", type: .stringClosure" : nil
 
-        "RubyCommand.Argument(name: \"#{name}\", value: #{sanitized_name}#{type_string})"
+        method_call = unless type_override == :string_callback || !is_optional 
+                        ".get()"
+                      else
+                        nil
+                      end
```
Comment (rogerluan) [inline fastlane/lib/fastlane/swift_fastlane_function.rb:291]: I think these double negatives are hard to follow 🙈 Perhaps inverting them will make it more easily readable, such as:

```suggestion
        method_call = if type_override != :string_callback || is_optional
                        ".get()"
                      else
                        nil
                      end 
```

note: I'm not rubyist so not sure if my logic is correct (e.g. idk if negating `==` in this case is the equivalent of `!=`)

**Review by rogerluan — COMMENTED** [review]
Mostly formatting suggestions 😁 but overall this is looking really good!

**Comment by minuscorp** [comment]
Yeah, don't take it so serious at the moment, still a work in progress in terms of code generation and so on, needs tons of code cleanup as you have commented (just had to change lots of autogenerated code by the compiler you know😪) about the naming, it indicates if the value of an item correspond to a `current` value or a user defined one (aka `new`) but any other naming is ok for me", like `default` `userDefined` and `nil` for example. I'm open to suggestions! ❤️

File: fastlane/lib/fastlane/swift_fastlane_function.rb (hunk 1 above)
Comment (minuscorp) [reply fastlane/lib/fastlane/swift_fastlane_function.rb:291]: Yeah, got rubocop offenses for my double negation logic, but needed for fast testing the possible states. I'll try yours!

File: fastlane/swift/ConfigItem.swift [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    case `nil`
+
+    public func get() -> T {
+        switch self {
+        case let .current(value):
+            return value
+        case let .new(value):
+            return value
+        case .nil:
+            fatalError("\(#function) getting called in a nil value")
```
[the remark this answers is not in the document]
Comment (minuscorp) [reply fastlane/swift/ConfigItem.swift:25]: I'll apply a filter onto current and nil values to the array of Ruby arguments, so it is really a dead path in theory, that's why I didn't wasted much time on the message but yours is much descriptive

**Comment by rogerluan** [comment]
Ahh got it!
I think `rubyDefaultValue` or `fastlaneDefault` would be easier to understand. And then `userDefined` and `nil` 💯

**Comment by ainame** [comment]
I'm keen to review Fastlane.swift related PRs but I've never really tried it🙂 Would you mind if you guide me on how Fastlane.swift works roughly? 🙏

**Comment by minuscorp** [comment]
You can checkout the docs for a quick start. Roughly this is a Swift runloop that attaches to a socket that it's opened by the Ruby side where the commands (which action to run and which arguments to be run with) are sent by the Swift internal interface so users just use autogenerated code that creates the messages for the socket. Ping me in slack if you need more info about it!❤️

**Comment by ainame** [comment]
Thanks a lot! I thought it is using something like dynamic callable😂 Will learn fastlane.swift gradually 🚶

**Comment by minuscorp** [comment]
It uses a bit of dark magic like getting the lane names from the method naming in the Fastfile class using `NSObject` and objc call table to run the correct method from the argument lane name.

File: fastlane/swift/ConfigItem.swift [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+//  If you want to add extra functionality to this project, create a new file in a
+//  new group so that it won't be marked for upgrade
+//
+
+import Foundation
+
+public enum ConfigItem<T> {
+    case fastlaneDefault(T)
+    case userDefined(T)
+    case `nil`
```
Comment (ainame) [inline fastlane/swift/ConfigItem.swift:16]: I don't fully understand this yet but could it be possible that we use only two cases; `fastlaneDefault` and `userDefined`?

File: fastlane/lib/fastlane/swift_fastlane_function.rb [hunk 4]
```diff
@@ -190,7 +193,13 @@ def parameters
         if default_value.nil?
           "#{param}: #{type}"
         else
-          "#{param}: #{type} = #{default_value}"
+          if type == "((String) -> Void)?"
+            "#{param}: #{type} = nil"
+          elsif optional
+            "#{param}: ConfigItem<#{type}> = .fastlaneDefault(#{default_value})"
+          else
+            "#{param}: #{type} = #{default_value}"
```
Comment (ainame) [inline fastlane/lib/fastlane/swift_fastlane_function.rb:201]: Why some params are treated as ConfigItem others are not? Wouldn't it be natural that we always use ConfigItem for every single param for actions?

File: fastlane/swift/ConfigItem.swift [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    func asRubyArgument(name: String, type: RubyCommand.Argument.ArgType? = nil) -> RubyCommand.Argument? {
+        if case let .userDefined(value) = self {
+            return RubyCommand.Argument(name: name, value: value, type: type)
+        }
```
Comment (ainame) [inline fastlane/swift/ConfigItem.swift:21]: Why `.fastlaneDefault(T)` is ignored here? Is that for document purpose?

**Review by ainame — COMMENTED** [review]
I had basic questions to help me understand this PR. Feel free to answer/ignore them😄 At least I understand that this can solve the issue that occurss due to default values being `nil` are given to the Ruby side.

File: fastlane/swift/ConfigItem.swift (hunk 3 above)
Comment (minuscorp) [reply fastlane/swift/ConfigItem.swift:16]: `case nil` serves the purpose of filling the case where the user defines as `nil` on-hand but that shouldn't be sent to the Ruby side

File: fastlane/lib/fastlane/swift_fastlane_function.rb (hunk 4 above)
Comment (minuscorp) [reply fastlane/lib/fastlane/swift_fastlane_function.rb:201]: `ConfigItem` is just designed for optional Ruby values

File: fastlane/swift/ConfigItem.swift (hunk 5 above)
Comment (minuscorp) [reply fastlane/swift/ConfigItem.swift:21]: We don't generate an argument when the case is not `.userDefined`, so `nil` will be returned and cleaned up in the `compactMap { $0 }` call before being sent into the socket.

File: fastlane/swift/ConfigItem.swift (hunk 3 above)
Comment (minuscorp) [reply fastlane/swift/ConfigItem.swift:16]: And as the extensions are being set over concrete non-optional values, we have to set at some point the `ExpressibleByNilLiteral` to be set to that extra `nil` case.

File: fastlane/swift/ConfigItem.swift [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+//
+//  ** NOTE **
+//  This file is provided by fastlane and WILL be overwritten in future updates
+//  If you want to add extra functionality to this project, create a new file in a
+//  new group so that it won't be marked for upgrade
+//
+
+import Foundation
+
+public enum ConfigItem<T> {
```
Comment (joshdholtz) [inline fastlane/swift/ConfigItem.swift:13]: A small nit pick maybe but thoughts on renaming this to `ConfigValue`? 😇 A config item is something a little more descriptive on the Ruby side where this is specifically just the value 🤷‍♂️

**Review by joshdholtz — CHANGES_REQUESTED** [review]
@minuscorp I tried running this with a command with `adb`. It has an `ConfigItem<String?>` and a `ConfigItem<String>` in it and I got some compile errors when trying to simply use Strings 👇 

Screen Shot 2021-05-04 at 4 10 39 PM

I solved these by adding the `.userDefined` but I don't think that is what the intended solution is, right? 🤷‍♂️ 

Screen Shot 2021-05-04 at 4 12 09 PM

It seems like the issue is with `ConfigItem<String>` as there is not a string literal (???) or something for it to auto convert/cast?

**Comment by minuscorp** [comment]
The issue is that for some reason the parameter is declared as optional but the real parameter later on is not `String?` so that might be causing the issue, I'll check that ASAP

File: fastlane/swift/ConfigItem.swift (hunk 6 above)
Comment (ainame) [reply fastlane/swift/ConfigItem.swift:13]: I agree with renaming. This is a bit confusing to me. Since it handles some sort of optionality on Swift side, `OptionalConfigValue` might make sense slightly more? 😆  
https://github.com/fastlane/fastlane/pull/18544#discussion_r625198720

Comment (minuscorp) [reply fastlane/swift/ConfigItem.swift]: I ended up renaming as @ainame suggested if that's fine with y'all 😄

**Comment by minuscorp** [comment]
I think I fixed those edge cases with the interpretation of what is an optional value and has a default value 🎉

**Review by joshdholtz — APPROVED** [review]
I think this is good to go now! I tested this out a bunch of times and reviewed some of the diff in the generated file 😊 I have no idea how you thought of this and made this because its so meta but 🏆
